heron/instance/src/java/com/twitter/heron/instance/Slave.java

```diff
@@ -170,10 +178,23 @@ public void run() {
     slaveLooper.loop();
   }
 
-  private void startInstance() {
-    instance.start();
-    isInstanceStarted = true;
-    LOG.info("Started instance.");
+  private void startInstanceIfNeeded() {
+    // To start the instance when:
+    //  1. We got the PhysicalPlan
+    //  2. The TopologyState == RUNNING
+    //  3. - If the topology is stateful and we got the stateful processing start signal
+    //     - If the topology is not stateful
+    if (helper != null && helper.getTopologyState().equals(TopologyAPI.TopologyState.RUNNING)) {
+      Map<String, Object> config = helper.getTopologyContext().getTopologyConfig();
+      boolean isTopologyStateful = String.valueOf(Config.TopologyReliabilityMode.EXACTLY_ONCE)
+          .equals(config.get(Config.TOPOLOGY_RELIABILITY_MODE));
+
+      if (!isTopologyStateful || isStatefulProcessingStarted) {
+        instance.start(instanceState);
```

Comment:
If the topology is not a stateful one, then a `null` state won't affect anything. It's not used by user code at all.

If the topology is a stateful one, and we haven't received the RestoreInstanceStateRequest. Then the components will start with a null value at first. But since the stateful topology wont start processing tuples until receives the `StartStatefulProcessingRequest`, so a null state won't break anything for the first. 

We just need to make sure that, there's always a `RestoreInstanceStateRequest` sent before the `StartStatefulProcessingRequest`. We can guarantee that the `instanceState` is properly initialized when the topology starts to processing